### ISSUE TITLE
Autocomplete slow to update with large data sets after material beta 12

### ISSUE THREAD

**Issue by micahwood** [angular/material2#7749]
#### Bug, feature request, or proposal:
When displaying an autocomplete list with a large number of options it is slow to display. It appears that this is caused by 2 event listeners being created for each item in the list. This slowness started after updating to beta 12.

You can see the large number of event listeners being set up if you open the Chrome dev tools and set the logging level to verbose.
`[Violation] Added non-passive event listener to a scroll-blocking 'touchstart' event. Consider marking event handler as 'passive' to make the page more responsive. See https://www.chromestatus.com/feature/5745543795965952`

#### What is the expected behavior?
Similar performance when using autocomplete with 200+ data sets to older versions of material.

#### What is the current behavior?
Rendering an autocomplete with a large number of options has a delay, it is even longer when the dev tools are open due to the large number of event listeners being created causing logs to be written to the console

#### What are the steps to reproduce?
Open dev tools, set console logging to verbose and click the "Click to toggle autocomplete" button .
You can also see the behavior by filtering the list, type anything to shorten the length of the list, then clear the input and new event listeners will be created again for each item in the list causing additional slowdown.
https://plnkr.co/edit/QhlHy7FrlvQbDeRfncCo?p=preview

#### What is the use-case or motivation for changing an existing behavior?
Improve performance

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
Angular: 4.4.4
Material: beta.12
All browsers

**Comment by radoslavpetranov** [comment]
I just spent 2 hours trying to figure out if I broke something performance-related somewhere after updating from material `beta.8` to `beta.12`. I was comparing the performance between the app on prod and the newly updated dev version and my country autocomplete is definitely causing noticeable lag in `beta.12`.

I have some form arrays where each internal form object contains a `country` autocomplete (so a whole bunch of options) and adding a few of these autocompletes is definitely causing pretty severe lag.

I also see thousands upon thousands of these messages

```
[Violation] Added non-passive event listener to a scroll-blocking 'touchstart' event. Consider marking event handler as 'passive' to make the page more responsive. See https://www.chromestatus.com/feature/5745543795965952
```

**Comment by dushkostanoeski** [comment]
Will this be addressed in the next beta?

**Comment by johnchristopherjones** [comment]
#5113 seems related.  I didn't see this problem in beta.10 but it appears to be quite nasty in beta.12.

**Comment by eqbalsajadi** [comment]
+1

**Comment by kylecordes** [comment]
If you're willing to load the numerous options on demand (rather than trying to put them all in the UI at once), you can get excellent results with unlimited size lists. Code that does so, and demo, link below.

https://github.com/OasisDigital/angular-material-obs-autocomplete

https://oasisdigital.github.io/angular-material-obs-autocomplete/

**Comment by mikeomeara1** [comment]
Hey guys, hate to "me too" but this is really become a problem and our users are clamoring...any idea when this might get looked at? 

We have some Autocomplete's that are taking 10+ seconds to initialize (which wasn't happening ~Beta12). We're at a point where we’re facing having to yank Autocomplete if we can’t get this fixed.

As always, appreciate all of your hard work and please let me know if there is anything we can do to assist.

**Comment by crisbeto** [comment]
@mikeomeara1 do you know approximately how many options you're rendering inside the autocomplete for it to take so long?

**Comment by mikeomeara1** [comment]
@crisbeto Generally speaking < 1000 works OK. Anything more than that there is anywhere from a perceptible "screen freeze" to a full 5-10 second gap while it loads up. Obviously the more you add the worse it gets. 

We didn't notice this issue (even with our huge lists) pre v5.0.0.

**Comment by crisbeto** [comment]
At what point are you running into the freeze? When loading up the view or when attempting to open the autocomplete panel?

**Comment by mikeomeara1** [comment]
On first showing the autocomplete, the delay is before the autocomplete panel is open (and once it's un-stuck, opening the panel is quick). However we're also seeing the same (or more) delay when we manually re-trigger the autocomplete panel after the user makes their first selection. 

I can send you a video if there is a way for me to send it to you directly (e.g. not have to post it)

**Comment by crisbeto** [comment]
A video might not be great for debugging. [I put together this demo for testing things out](https://stackblitz.com/edit/angular-material2-issue-getlqt) and there definitely is a noticeable delay both when rendering the autocomplete and when opening it. Are you seeing something similar in your case? 

As for improving the performance, we can relatively easily add the ability to have your `mat-autocomplete` be attached to an `ng-template` which will help with the initial load, because it won't render the options when they aren't being shown. Improving the opening of the panel is a different story, because we still have 2000+ components that need to be instantiated.

**Comment by mikeomeara1** [comment]
Fair enough. OK so, from what I can see here, this is running MUCH faster than what I'm seeing...this is how it USED to work. Even when I up the items to 10,000...it works how I would expect. So, that means one of two things:

1. Something got fixed between v5.0.0 and v5.0.1
2. I have something else broken here that happened when we upgraded from Beta.12 to 5.0.0

I'll try and upgrade here soon and report back if it solves the problem.

**Comment by crisbeto** [comment]
I don't think much has changed in the autocomplete between the 5.0.x version for it to have any impact on performance. My guess is that your template might be a bit more complicated than my demo.

**Comment by mikeomeara1** [comment]
It's almost exactly the same as yours actually...just with an `onSelectionChange` and async pipe.
[https://gist.github.com/mikeomeara1/facf01de71bed795b49457ae8bfdde23](https://gist.github.com/mikeomeara1/facf01de71bed795b49457ae8bfdde23)

 I just pulled your demo into my app and it's doing the same thing...taking a VERY long time to load up and reopen. Curious.

**Comment by apoorv-techelogy** [comment]
waiting for it to get resolved as soon as possible

**Comment by mikeomeara1** [comment]
@crisbeto I have some good news. Upgrading from v5.0.0 to v5.1.1 solved this problem. All of our autocomplete and select's are now working as we would expect (and the delays when rendering or opening have gone from 8-10 seconds down to <1 second again).

So, whatever was done during that time seemed to resolve the issue (at least for us). 

As usual, truly appreciate your help.

**Comment by Swoox** [comment]
@mikeomeara1 cheers this solves the issues I had.

**Comment by EvaristoPerez** [comment]
@mikeomeara1 thanks form my side too! big help. Now @crisbeto I think the changes that fixes this are more related to 5.1.0 than 5.1.1 that is why maybe you did not find relationships. So there at 5.0.1 on the ripple side it is mentioned that they fixed to optionally register those events mentioned in the original problem from @micahwood . Anyways thanks evryone big help for me though.

**Comment by DibyodyutiMondal** [comment]
I would like +1 to this.

In my setup, I'm using angular 16 with signals

I have a an autocomplete for selecting countries.
The countries are simply plain objects with `name`, `isoCode` and `flag` properties.
I use the `[displayWith]` attribute as per the docs, as well as the `trackBy` for the `mat-option`s, but these have not affected anything at all


The first warnings I noticed warnings on the console were like so:
```
[Violation] Forced reflow while executing JavaScript took 30ms
```
This warning occurs whenever I click the input, which opens the dropdown panel. 30ms is barely anything, so this is not really a problem.

If I do not type anything, and simply select from the list, the selection is fast and performant. No warnings are issued.

But, if I do even a single keypress, which filters the options, and then select any of the filtered options, then there is a noticeable delay:
```
[Violation] 'keydown' handler took 437ms
```
If I use the mouse, it tells me that the `click` handler took ~400-500ms. During that delay, the panel stays open, and then finally closes, at which point the input also updates with the selected value.

I hope this information will be useful in figuring out where the problem lies.
